go/service/rekey_ui_handler.go

```diff
@@ -0,0 +1,253 @@
+package service
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"golang.org/x/net/context"
+
+	"github.com/keybase/client/go/libkb"
+	keybase1 "github.com/keybase/client/go/protocol"
+	"github.com/keybase/gregor"
+)
+
+const rekeyHandlerName = "RekeyUIHandler"
+
+type RekeyUIHandler struct {
+	libkb.Contextified
+	connID         libkb.ConnectionID
+	alwaysAlive    bool
+	updatersMu     sync.RWMutex
+	updaters       map[int]*rekeyStatusUpdater
+	notifyComplete chan int // for testing purposes
+	scorer         func(g *libkb.GlobalContext, existing keybase1.ProblemSet) (keybase1.ProblemSet, error)
+}
+
+var _ libkb.GregorInBandMessageHandler = (*RekeyUIHandler)(nil)
+
+func NewRekeyUIHandler(g *libkb.GlobalContext, connID libkb.ConnectionID) *RekeyUIHandler {
+	return &RekeyUIHandler{
+		Contextified: libkb.NewContextified(g),
+		connID:       connID,
+		updaters:     make(map[int]*rekeyStatusUpdater),
+		scorer:       scoreProblemFolders,
+	}
+}
+
+func (r *RekeyUIHandler) Create(ctx context.Context, category string, item gregor.Item) (bool, error) {
+	switch category {
+	case "kbfs_tlf_rekey_needed":
+		return true, r.rekeyNeeded(ctx, item)
+	default:
+		return false, fmt.Errorf("unknown RekeyUIHandler category: %q", category)
+	}
+}
+
+func (r *RekeyUIHandler) Dismiss(ctx context.Context, category string, item gregor.Item) (bool, error) {
+	return false, nil
+}
+
+func (r *RekeyUIHandler) IsAlive() bool {
+	return true
+}
+
+func (r *RekeyUIHandler) Name() string {
+	return rekeyHandlerName
+}
+
+func (r *RekeyUIHandler) rekeyNeeded(ctx context.Context, item gregor.Item) error {
+	if item.Body() == nil {
+		return errors.New("gregor handler for kbfs_tlf_rekey_needed: nil message body")
+	}
+
+	var problemSet keybase1.ProblemSet
+	if err := json.Unmarshal(item.Body().Bytes(), &problemSet); err != nil {
+		r.G().Log.Debug("error: %s", err)
+		return err
+	}
+
+	if r.G().Clock.Now().Sub(item.Metadata().CTime()) > 10*time.Second {
+		// if the message isn't fresh, get:
+		var err error
+		problemSet, err = r.scorer(r.G(), problemSet)
+		if err != nil {
+			return err
+		}
+	}
+
+	// if the tlf list is empty, dismiss the gregor notification
+	if len(problemSet.Tlfs) == 0 {
+		r.G().Log.Debug("problem set tlf list empty, dismissing gregor notification")
+		return r.G().GregorDismisser.DismissItem(item.Metadata().MsgID())
+	}
+
+	// get the rekeyUI
+	rekeyUI, sessionID, err := r.G().UIRouter.GetRekeyUI()
+	if err != nil {
+		r.G().Log.Errorf("failed to get RekeyUI: %s", err)
+		return err
+	}
+	if rekeyUI == nil {
+		r.G().Log.Error("got nil RekeyUI")
+		return errors.New("got nil RekeyUI")
```

Comment:
I wonder if we should dump an Error to the service log in this case, since we do want them to know.  So in other words, can we have a RekeyUIHandler attached even if running in CLI mode?